### PR TITLE
Tor lazy download

### PR DESCRIPTION
**Pull request by darkdh**
Resolves https://github.com/brave/brave-browser/issues/9177

This PR will download tor executable only when first Tor window is opened and ~show indicator and block URL bar~ defer requests until Tor process is launched. Tor executable will not be deleted after relaunch, it will only be deleted when tor is disabled in settings or by group policy

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [x] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [x] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

### New profile and relaunch
1. Launch Brave with clean profile
2. Go to brave://components and check there is no "Tor Client Updater"
3. Open tor window
4. Go to https://check.torproject.org/
5. It will show request pending like
Screen Shot 2020-07-20 at 22 27 57

6. After tor process is launched,  it will get https://check.torproject.org/ loaded eventually
7.  "Tor Client Updater" shows up in brave://components
8. It should be able connecting to tor network
9. Close tor window and repeat step 3-8
10. Relaunch Brave
11.  There is no "Tor Client Updater" in brave://components
12. Go to user folder and check tor executable component is still there 
(`cldoidikboihgcjfkhdeidbpclkineef`/MacOS, `biahpgbdmdkfgndcmfiipgcebobojjkp`/Linux, `cpoalefficncklhjfpglfiplenlpccdb`/Windows)
13. Repeat step 3-8

### Emulate slow network
1. Launch Brave with clean profile
2. Turn off internet access for Brave
3. Open tor window and go to https://check.torproject.org/
4. We will see request is pending like
Screen Shot 2020-07-20 at 22 27 57

5. Turn internet back on
6. Go to brave://components and manually update "Tor Client Updater" on normal window
7. Go back to tor window and wait for tor process launched
8. It will get https://check.torproject.org/ loaded eventually

### Disable Tor by settings
1. Launch Brave with clean profile
2. Open tor window
3. Wait for tor process launched
4. Close Tor window
5. Go to brave://settings/extensions
6. Turn toggle `Private Window with Tor` off
7. Go to user folder and tor executable component should be gone
(`cldoidikboihgcjfkhdeidbpclkineef`/MacOS, `biahpgbdmdkfgndcmfiipgcebobojjkp`/Linux, `cpoalefficncklhjfpglfiplenlpccdb`/Windows)

### Disable Tor by group policy (Windows)
1. Launch Brave with clean profile
2. Open tor window
3. Wait for tor process launched
4. Exit Brave
5. Open regedit
6. For 64 bit, create the following keys:
`BraveSoftware\Brave`
under the existing key:
`HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Policies\`
(for 32 bit, you can edit `HKEY_LOCAL_MACHINE\SOFTWARE\Policies\`)
Create a DWORD value at this path called `TorDisabled` and set the value as 1
7. Launch Brave
8. Go to user folder and tor executable component should be gone
(`cpoalefficncklhjfpglfiplenlpccdb`/Windows)

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: browser/extensions/brave_extension_management.cc [hunk 1]
```diff
@@ -69,4 +63,11 @@ void BraveExtensionManagement::OnExtensionUnloaded(
     pref_service_->SetBoolean(kIPFSCompanionEnabled, false);
 }
 
+void BraveExtensionManagement::OnTorDisabledChanged() {
```
Comment (simonhong) [inline browser/extensions/brave_extension_management.cc:66]: I think tor executables will be deleted immediately after user disables tor option from settings page.

Comment (simonhong) [reply browser/extensions/brave_extension_management.cc:66]: Ah, sorry I didn't see STR. This is expected behavior.
Doesn't this make any issue if deletion happens while tor window is opened?

Comment (darkdh) [reply browser/extensions/brave_extension_management.cc:66]: It won't have any issues when tor window is opened, it's already loaded into memory.

Comment (simonhong) [reply browser/extensions/brave_extension_management.cc:66]: got it.

File: browser/extensions/brave_tor_client_updater.cc [hunk 2]
```diff
@@ -135,14 +135,20 @@ void BraveTorClientUpdater::Register() {
...
 
+void BraveTorClientUpdater::Unregister() {
+  // We don't call BraveComponent::Unregister here in order to prevent tor
+  // executable component from getting deleted when last tor window closed
+  registered_ = false;
+}
+
 void BraveTorClientUpdater::Cleanup() {
-  // Delete tor binaries if tor is disabled by gpo.
+  // Delete tor binaries if tor is disabled.
```
Comment (simonhong) [inline browser/extensions/brave_tor_client_updater.cc:145]: I think unregistering tor component will delete component folder immediately?

Comment (darkdh) [reply browser/extensions/brave_tor_client_updater.cc:145]: yes, but it won't work for these scenario:
1. User once opened tor window and get tor executable downloaded
2. User relaunch browser
3. User disable tor in setting

Comment (simonhong) [reply browser/extensions/brave_tor_client_updater.cc:145]: Sorry, I don't understand. how unregistering component is related with above scenario?

Comment (darkdh) [reply browser/extensions/brave_tor_client_updater.cc:145]: now registration only happens with "Tor Window Opened",  after relaunch we won't register the component until tor window opened

Comment (simonhong) [reply browser/extensions/brave_tor_client_updater.cc:145]: I thought deleting directory directly and unregistering component would have same result.

Comment (darkdh) [reply browser/extensions/brave_tor_client_updater.cc:145]: After relaunch, if we don't open tor window, then it is in an unregistered state. Then unregister it won't have any effects.

Comment (darkdh) [reply browser/extensions/brave_tor_client_updater.cc:145]: https://github.com/brave/brave-core/pull/6081/commits/ec6adc21f3882bd9885877f83844911a35ef5b48#diff-6bf7995a812906ed34d369d71cfa4de3R50
This is where we register tor client updater

File: browser/extensions/brave_extension_management.h [hunk 3]
```diff
@@ -8,6 +8,7 @@
 
 #include "base/scoped_observer.h"
 #include "chrome/browser/extensions/extension_management.h"
+#include "components/prefs/pref_change_registrar.h"
```
Comment (bridiver) [inline browser/extensions/brave_extension_management.h:11]: do we need to add a dep for this?

**Comment by bsclifton** [comment]
Nice work, @darkdh 😄

File: browser/extensions/brave_extension_management.h (hunk 3 above)
Comment (darkdh) [reply browser/extensions/brave_extension_management.h:11]: no, deps for `//components/prefs` is already specified in `browser/extensions/BUILD.gn`

**Comment by darkdh** [comment]
follow up issue created 
https://github.com/brave/brave-browser/issues/10867

File: chromium_src/chrome/browser/profiles/profile_window.cc [hunk 4]
```diff
@@ -42,6 +44,10 @@ void OpenBrowserWindowForTorProfile(ProfileManager::CreateCallback callback,
   profiles::OpenBrowserWindowForProfile(
       callback, always_create, is_new_profile, unblock_extensions,
       profile->GetOffTheRecordProfile(), status);
+  tor::TorProfileService* service =
+    TorProfileServiceFactory::GetForProfile(profile);
```
Comment (bridiver) [inline chromium_src/chrome/browser/profiles/profile_window.cc:48]: lint nit - 4 spaces for continuation of a line

File: browser/tor/tor_tab_helper.cc [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents->GetBrowserContext());
+  if (!brave::IsTorProfile(profile))
+    return;
+  TorTabHelper::CreateForWebContents(web_contents);
+}
+
+void TorTabHelper::DidFinishNavigation(
+    content::NavigationHandle* navigation_handle) {
+  if (navigation_handle->GetNetErrorCode() != net::ERR_PROXY_CONNECTION_FAILED)
```
Comment (bridiver) [inline browser/tor/tor_tab_helper.cc:32]: a comment explaining this would probably be useful

Comment (darkdh) [reply browser/tor/tor_tab_helper.cc:35]: fixed

File: chromium_src/chrome/browser/profiles/profile_window.cc (hunk 4 above)
Comment (darkdh) [reply chromium_src/chrome/browser/profiles/profile_window.cc]: fixed

**Comment by darkdh** [comment]
Unrelated CI failure
```
16:28:48  1 test failed:
16:28:48      RewardsBrowserTest.NotVerifiedWallet (../../brave/components/brave_rewards/browser/test/rewards_browsertest.cc:241)
```
https://ci.brave.com/job/pr-brave-browser-tor-lazy-download/22/execution/node/309/log/

**Pull request merged**